### PR TITLE
[Draft] Long Context Training VRAM Optimization

### PR DESCRIPTION
**Pull request by jiapingW**
## Motivation

This RP handle the VRAM in 64k training with sp=8 and decreae it from 94G per GPU to 76G per GPU. And this contains the PR https://github.com/sgl-project/SpecForge/pull/429. We have do these things in this PR:
- Optimize long context training VRAM usage
- Support handle comprehensive `tool-use` format message

## Accuracy Test

We have test it on private task to verify its correctness. We can now train offline on H20 with more than 128k data.
## Checklist

- [✅] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

File: specforge/core/eagle3.py [hunk 1]
```diff
@@ -206,8 +207,12 @@ def forward(
...
 
             # Step 5.4: get logits
-            logits = self.draft_model.compute_logits(hidden_states)
-            logits = gather_outputs_and_unpad(logits, gather_dim=1)
+            logits_ = self.draft_model.compute_logits(hidden_states)
+            # from .forkedpdb import ForkedPdb
+            # ForkedPdb().set_trace()
+            logits = gather_outputs_and_unpad(logits_, gather_dim=1)
+            del logits_
+            torch.cuda.empty_cache()
```
[the remark this answers is not in the document]
Comment (FrankLeeeee) [reply specforge/core/eagle3.py]: Same question as gemini, how much impact will this have on the training performance?

File: scripts/train_eagle3.py [hunk 2]
```diff
@@ -589,14 +596,16 @@ def run_forward(
...
-            target = target_model(data["target"].cuda())
             input_ids, target, loss_mask = target_model.preprocess(
-                input_ids, target, loss_mask
+                data["input_ids"], data["target"], data["loss_mask"]
             )
+            input_ids = input_ids.cuda()
+            target = target_model(
+                target.cuda()
+            )  # The `data['target']` value occupies a large amount of GPU memory, with a shape of [seqlen, vocab_size]. It needs to be processed before being loaded into the GPU.
+            loss_mask = loss_mask.cuda()
```
Comment (FrankLeeeee) [inline scripts/train_eagle3.py:608]: What is the impact of this on performance? if it is large, maybe we can set it as a flag to control whether do this on GPU or CPU.

Comment (jiapingW) [reply scripts/train_eagle3.py:608]: We can simply compute this: vocab size(150000) * seq_length(64k) will cost 10G more.

Comment (jiapingW) [reply scripts/train_eagle3.py:608]: The reason is target_head's `preprocess` function will use `padding` will generate an extra copy of the `target` memory.

File: specforge/core/eagle3.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jiapingW) [reply specforge/core/eagle3.py]: I‘ve deleted it  and the impact on time is minimal based on the current results.

File: scripts/train_eagle3.py (hunk 2 above)
Comment (uygnef) [reply scripts/train_eagle3.py:608]: we can split hidden state in dataset __getitem__ for usp to reduce memory use.

Comment (jiapingW) [reply scripts/train_eagle3.py:608]: Yes, I think this is a better optimization method. Can you help add this optimization?

> we can split hidden state in dataset **getitem** for usp to reduce memory use.

Comment (uygnef) [reply scripts/train_eagle3.py:608]: https://github.com/sgl-project/SpecForge/pull/454 
Hi, I've finished the updates. Note that SP currently works with batch size 1. This seems reasonable for long-sequence scenarios to avoid OOM, but I'm open to feedback. Ready for review!

Comment (jiapingW) [reply scripts/train_eagle3.py:608]: Great, I'll help review it.

**Comment by ggg-s** [comment]
hi @jiapingW  Does "128k data" refer to a training sequence length of 128k?

**Comment by jiapingW** [comment]
> hi @jiapingW Does "128k data" refer to a training sequence length of 128k?

Yes.
